### PR TITLE
feat: Firestore datastore mode index support

### PR DESCRIPTION
**Pull request by pl04351820**
feat: Firestore datastore mode index support

This includes a support for datastore mode index under Firestore index surface. This allows Terraform users to manage datastore mode index for non-default databases via Firestore index surface.

If this PR is for Terraform, I acknowledge that I have:

- [X] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [X] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/third_party/terraform/tests) (for handwritten resources or update tests).
- [X] [Generated Terraform providers](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/docs/content/docs/getting-started/generate-providers.md), and ran [`make test` and `make lint`](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/run-provider-tests/#run-unit-tests) in the generated providers to ensure it passes unit and linter tests.
- [X] [Ran](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/docs/content/develop/run-tests.md) relevant acceptance tests using my own Google Cloud project and credentials (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [X] Read [Write release notes](https://googlecloudplatform.github.io/magic-modules/contribute/release-notes/) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
firestore: added `api_scope` field to `google_firestore_index` resource
```

### CODE REVIEW

**Review by rwhogg — APPROVED** [review]

File: mmv1/products/firestore/Index.yaml
Comment (shuyama1) [inline mmv1/products/firestore/Index.yaml:108]: Question: why do we need to have `ignore_read` for this field? Does API not return values for this field?

File: mmv1/templates/terraform/examples/firestore_index_datastore_mode.tf.erb
Comment (shuyama1) [inline mmv1/templates/terraform/examples/firestore_index_datastore_mode.tf.erb:1]: We need to register this example in `Index.yaml` to be able to generate test and doc. See step 5 at https://googlecloudplatform.github.io/magic-modules/develop/test/#add-a-create-test

File: mmv1/products/firestore/Index.yaml
Comment (shuyama1) [inline mmv1/products/firestore/Index.yaml:100]: is this field updatable? Is so, can we add/modify an update test to test it? See details at https://googlecloudplatform.github.io/magic-modules/develop/test/#add-an-update-test

**Review by shuyama1 — CHANGES_REQUESTED** [review]

File: mmv1/products/firestore/Index.yaml
Comment (pl04351820) [reply mmv1/products/firestore/Index.yaml:109]: No. Firestore does not support `index update` operation.

File: mmv1/templates/terraform/examples/firestore_index_datastore_mode.tf.erb
Comment (pl04351820) [reply mmv1/templates/terraform/examples/firestore_index_datastore_mode.tf.erb:1]: Done. Thanks for the pointer!

File: mmv1/products/firestore/Index.yaml
Comment (pl04351820) [reply mmv1/products/firestore/Index.yaml:117]: This is really for compatibility (though unclear to me what's the best way to do this in Terraform).

The goal is making ANY_API as default value (and hide if it's default). Only show API_SCOPE if it's DATASTORE_MODE_API. Please let me know if there is a better way to achieve this.

Comment (shuyama1) [reply mmv1/products/firestore/Index.yaml:117]: I've seen you add a terraform side default `default_value: :ANY_API` already, which means that if users do not specify this fields, terraform will send `ANY_API`. Is `ANY_API` also the service side default for the field?
Plus, will API return the value back if users specify either `ANY_API` or `DATASTORE_MODE_API` for the field. 
If so, we'd remove `ignore_read` as this will prevent from reading back from the API.

**Review by shuyama1 — COMMENTED** [review]
Left comment inline in https://github.com/GoogleCloudPlatform/magic-modules/pull/8713#discussion_r1304685730 incase you're not aware of. Sorry that github comment reply is not so easy to be spotted.

File: mmv1/products/firestore/Index.yaml
Comment (pl04351820) [reply mmv1/products/firestore/Index.yaml]: SG. Thanks for the suggestions!

File: mmv1/.ruby-version
Comment (rwhogg) [inline mmv1/.ruby-version:1]: Can you remove this change?

**Review by shuyama1 — COMMENTED** [review]
Overall LGTM, only one small comment. Please remove the change as @rwhogg mentioned. (uncheck the [commit](https://github.com/GoogleCloudPlatform/magic-modules/pull/8713/commits/0f4e72e5977574a1143eb6b21f205da4879726ca) from this PR would be nice). Thanks!

File: mmv1/products/firestore/Index.yaml
Comment (rwhogg) [inline mmv1/products/firestore/Index.yaml:64]: Can you also remove "with `type` set to `"FIRESTORE_NATIVE"`" bit above?

Comment (pl04351820) [reply mmv1/products/firestore/Index.yaml:62]: Done

**Review by shuyama1 — APPROVED** [review]
LGTM! Sorry we're currently locking the main branch contribution for major release preparation. I will merge this PR after it's unlocked, likely before early next week. Thanks!

**Comment by shuyama1** [comment]
unlocked now! rerun tests before merging! /gcbrun

**Comment by shuyama1** [comment]
@pl04351820 Would you mind rebasing the PR. I think the `terraform-google-conversion-test` failures will probably be resolved after rebasing

**Comment by pl04351820** [comment]
@shuyama1 Done!

**Comment by pl04351820** [comment]
Looks like TestAccContainerNodePool and TestAccContainerCluster still failed with unrelated issues

**Review by shuyama1 — APPROVED** [review]
The failing tests are unrelated

**Pull request merged**
